File: air-androidd/app/src/main/java/ru/alfabank/alfamir/favorites/presentation/favorite_profile/FavoriteProfilePresenter.java
package ru.alfabank.alfamir.favorites.presentation.favorite_profile;//package ru.alfabank.alfamir.favorites.presentation.favorite_profile;
//
//import com.google.common.base.Strings;
//
//import java.util.List;
//
//import javax.inject.Inject;
//
//import io.reactivex.android.schedulers.AndroidSchedulers;
//import io.reactivex.disposables.CompositeDisposable;
//import io.reactivex.schedulers.Schedulers;
//import ru.alfabank.alfamir.favorites.domain.usecase.GetFavoriteProfileList;
//import ru.alfabank.alfamir.favorites.presentation.dto.FavoriteProfile;
//import ru.alfabank.alfamir.favorites.presentation.favorite_profile.contract.FavoriteProfileAdapterContract;
//import ru.alfabank.alfamir.favorites.presentation.favorite_profile.contract.FavoriteProfileContract;
//import ru.alfabank.alfamir.favorites.presentation.favorite_profile.dummy_view.FavoriteProfileViewDummy;
//import ru.alfabank.alfamir.GetImage;
//import ru.alfabank.alfamir.image.presentation.dto.Image;
//import ru.alfabank.alfamir.utility.static_utilities.InitialsMaker;
//
//public class FavoriteProfilePresenter implements FavoriteProfileContract.Presenter {
//
//    private FavoriteProfileContract.View mView;
//    private FavoriteProfileAdapterContract.Adapter mAdapter;
//    private GetFavoriteProfileList mGetFavoriteProfileList;
//    private GetImage mGetImage;
//    private CompositeDisposable mCompositeDisposable = new CompositeDisposable();
//    private List<FavoriteProfile> mFavoriteProfiles;
//    private boolean mIsDataLoaded;
//
//
//    @Inject
//    FavoriteProfilePresenter(GetFavoriteProfileList getFavoriteProfileList,
//                             GetImage fillImage){
//        mGetFavoriteProfileList = getFavoriteProfileList;
//        mGetImage = fillImage;
//    }
//
//    private void loadFavoriteProfileList(){
//        mCompositeDisposable.add(mGetFavoriteProfileList.execute(new GetFavoriteProfileList.RequestValues())
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe(responseValue -> {
//                    mFavoriteProfiles = responseValue.getmFavoriteProfiles();
//                    getView().showFavoriteList();
//                    getView().hideLoadingProgress();
//                }, throwable -> {
//                    getView().hideLoadingProgress();
//                }));
//    }
//
//    @Override
//    public int getListSize() {
//        return mFavoriteProfiles.size();
//    }
//
//    @Override
//    public int getItemViewType(int position) {
//        return 0;
//    }
//
//    @Override
//    public void takeListAdapter(FavoriteProfileAdapterContract.Adapter adapter) {
//        mAdapter = adapter;
//    }
//
//    @Override
//    public FavoriteProfileAdapterContract.Adapter getAdapter() {
//        if (mAdapter == null) return new FavoriteProfileViewDummy();
//        return mAdapter;
//    }
//
//    @Override
//    public void takeView(FavoriteProfileContract.View view) {
//        mView = view;
////        if(mIsDataLoaded) return;
//        getView().showLoadingProgress();
//        loadFavoriteProfileList();
//    }
//
//    @Override
//    public void dropView() {
//        mView = null;
//        mAdapter = null;
//        mCompositeDisposable.dispose();
//    }
//
//    @Override
//    public FavoriteProfileContract.View getView() {
//        if (mView == null) return new FavoriteProfileViewDummy();
//        return mView;
//    }
//
//    @Override
//    public void bindListRowProfile(FavoriteProfileAdapterContract.ProfileRowView rowView, int position) {
//        FavoriteProfile favoriteProfile =  mFavoriteProfiles.get(position);
//        String email = favoriteProfile.getEmail();
//        String imageUrl = favoriteProfile.getImageUrl();
//        String login = favoriteProfile.getLogin();
//        String name = favoriteProfile.getName();
//        String title = favoriteProfile.getTitle();
//        String workPhone = favoriteProfile.getWorkPhone();
//        rowView.setName(name);
//        rowView.setTitle(title);
//        if(Strings.isNullOrEmpty(imageUrl)){
//            String initials = InitialsMaker.formInitials(name);
//            rowView.setPlaceHolder(initials);
//        } else {
//            mCompositeDisposable.add(mGetImage.execute(new GetImage.RequestValues(imageUrl, 80))
//                    .subscribeOn(Schedulers.io())
//                    .observeOn(AndroidSchedulers.mainThread())
//                    .subscribe(responseValue -> {
//                        Image image = responseValue.fillImage();
//                        String encodedImage = image.getEncodedImage();
//                        boolean isCached = image.getIsCached();
//                        if(Strings.isNullOrEmpty(encodedImage)) return; // TODO do i really need this?
//                        if(isCached){
//                            rowView.setImage(encodedImage, true);
//                        } else {
//                            getAdapter().onImageLoaded(position, 0, encodedImage, true);
//                        }
//                    }, throwable -> {
//
//                    }));
//        }
//    }
//
//    @Override
//    public void onItemClicked(int position) {
//        FavoriteProfile favoriteProfile =  mFavoriteProfiles.get(position);
//        String email = favoriteProfile.getEmail();
//        String imageUrl = favoriteProfile.getImageUrl();
//        String login = favoriteProfile.getLogin();
//        String name = favoriteProfile.getName();
//        String title = favoriteProfile.getTitle();
//        String workPhone = favoriteProfile.getWorkPhone();
//        mView.openProfileActivityUi(login);
//    }
//
//    @Override
//    public void onSearchClicked() {
//        mView.openSearchUi();
//    }
//}

File: air-androidd/app/src/main/java/ru/alfabank/alfamir/favorites/presentation/favorite_profile/FavoriteProfilePresenter.kt
package ru.alfabank.alfamir.favorites.presentation.favorite_profile

import android.annotation.SuppressLint
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import ru.alfabank.alfamir.favorites.domain.usecase.GetFavoriteProfileList
import ru.alfabank.alfamir.favorites.domain.usecase.RemoveFavoritePerson
import ru.alfabank.alfamir.favorites.presentation.dto.FavoriteProfile
import ru.alfabank.alfamir.favorites.presentation.favorite_profile.contract.FavoriteProfileAdapterContract
import ru.alfabank.alfamir.favorites.presentation.favorite_profile.contract.FavoriteProfileContract
import ru.alfabank.alfamir.favorites.presentation.favorite_profile.dummy_view.FavoriteProfileViewDummy
import ru.alfabank.alfamir.image.domain.usecase.GetImage
import ru.alfabank.alfamir.utility.static_utilities.InitialsMaker
import javax.inject.Inject

class FavoriteProfilePresenter
@Inject
internal constructor(private val mGetFavoriteProfileList: GetFavoriteProfileList,
                     private val mRemoveFavoritePerson: RemoveFavoritePerson,
                     private val mGetImage: GetImage) : FavoriteProfileContract.Presenter {

    private var mView: FavoriteProfileContract.View? = null
    private var mAdapter: FavoriteProfileAdapterContract.Adapter? = null
    private val mCompositeDisposable = CompositeDisposable()
    private var mFavoriteProfiles: MutableList<FavoriteProfile> = mutableListOf()
    private val job = Job()
    private val uiScope = CoroutineScope(Dispatchers.Main + job)

    private fun loadFavoriteProfileList() {
        mCompositeDisposable.add(mGetFavoriteProfileList.execute(GetFavoriteProfileList.RequestValues())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ responseValue ->
                    mFavoriteProfiles = responseValue.getmFavoriteProfiles()
                    view.showFavoriteList()
                    view.hideLoadingProgress()
                }, { view.hideLoadingProgress() }))
    }

    override fun getListSize(): Int {
        return mFavoriteProfiles.size
    }

    override fun getItemViewType(position: Int): Int {
        return 0
    }

    override fun takeListAdapter(adapter: FavoriteProfileAdapterContract.Adapter) {
        mAdapter = adapter
    }

    override fun getAdapter(): FavoriteProfileAdapterContract.Adapter {
        return if (mAdapter == null) FavoriteProfileViewDummy() else mAdapter!!
    }

    override fun takeView(view: FavoriteProfileContract.View) {
        mView = view
        getView().showLoadingProgress()
        loadFavoriteProfileList()
    }

    override fun dropView() {
        mView = null
        mAdapter = null
        mCompositeDisposable.dispose()
        job.cancel()
    }

    override fun getView(): FavoriteProfileContract.View {
        return if (mView == null) FavoriteProfileViewDummy() else mView!!
    }

    override fun bindListRowProfile(rowView: FavoriteProfileAdapterContract.ProfileRowView, position: Int) {
        val favoriteProfile = mFavoriteProfiles[position]
        val imageUrl = favoriteProfile.imageUrl
        val name = favoriteProfile.name
        val title = favoriteProfile.title
        rowView.setName(name)
        rowView.setTitle(title)
        if (imageUrl.isNullOrEmpty()) {
            val initials = InitialsMaker.formInitials(name)
            rowView.setPlaceHolder(initials)
        } else {

            uiScope.launch {
                mCompositeDisposable.add(mGetImage.bitmap(imageUrl, 80)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ rowView.setImage(it, true) }, Throwable::printStackTrace))
            }
        }
    }

    override fun onItemClicked(position: Int) {
        val favoriteProfile = mFavoriteProfiles[position]
        val login = favoriteProfile.login
        mView!!.openProfileActivityUi(login)
    }

    override fun onSearchClicked() {
        mView!!.openSearchUi()
    }

    @SuppressLint("CheckResult")
    override fun delete(position: Int) {
        val profile = mFavoriteProfiles[position]
        mRemoveFavoritePerson.execute(profile.login)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({}, {})
        mFavoriteProfiles.removeAt(position)

    }
}